pkg/storage/utils/decomposedfs/upload.go

```diff
@@ -478,535 +262,50 @@ func (fs *Decomposedfs) PurgeExpiredUploads(purgedChan chan<- tusd.FileInfo) err
-	discardBlob := func() {
-		if err := upload.fs.tp.DeleteBlob(newNode); err != nil {
-			sublog.Err(err).Str("blobid", newNode.BlobID).Msg("Decomposedfs: failed to discard blob in blobstore")
-		}
-	}
-
-	// lock metadata node
-	lock, err := filelocks.AcquireWriteLock(targetPath)
-	if err != nil {
-		discardBlob()
-		return errtypes.InternalError(err.Error())
-	}
-	releaseLock := func() {
-		// ReleaseLock returns nil if already unlocked
-		if err := filelocks.ReleaseLock(lock); err != nil {
-			sublog.Err(err).Msg("Decomposedfs:could not unlock node")
-		}
-	}
-	defer releaseLock()
-
-	// check if match header again as safeguard
-	var oldMtime time.Time
-	versionsPath := ""
-	if fi, err = os.Stat(targetPath); err == nil {
-		// When the if-match header was set we need to check if the
-		// etag still matches before finishing the upload.
-		if ifMatch, ok := upload.info.MetaData["if-match"]; ok {
-			var targetEtag string
-			targetEtag, err = node.CalculateEtag(newNode.ID, fi.ModTime())
-			if err != nil {
-				discardBlob()
-				return errtypes.InternalError(err.Error())
-			}
-			if ifMatch != targetEtag {
-				discardBlob()
-				return errtypes.Aborted("etag mismatch")
-			}
-		}
-
-		// versions are stored alongside the actual file, so a rename can be efficient and does not cross storage / partition boundaries
-		versionsPath = upload.fs.lu.InternalPath(spaceID, newNode.ID+node.RevisionIDDelimiter+fi.ModTime().UTC().Format(time.RFC3339Nano))
-
-		// remember mtime of existing file so we can apply it to the version
-		oldMtime = fi.ModTime()
-	}
-
-	// read metadata
-
-	// attributes that will change
-	attrs := map[string]string{
-		xattrs.BlobIDAttr:   newNode.BlobID,
-		xattrs.BlobsizeAttr: strconv.FormatInt(newNode.Blobsize, 10),
-
-		// update checksums
-		xattrs.ChecksumPrefix + "sha1":    string(sha1h.Sum(nil)),
-		xattrs.ChecksumPrefix + "md5":     string(md5h.Sum(nil)),
-		xattrs.ChecksumPrefix + "adler32": string(adler32h.Sum(nil)),
-	}
-
-	// create version node with current metadata
-
-	var newMtime time.Time
-	// if file already exists
-	if versionsPath != "" {
-		// touch version node
-		file, err := os.Create(versionsPath)
-		if err != nil {
-			discardBlob()
-			sublog.Err(err).Str("version", versionsPath).Msg("could not create version node")
-			return errtypes.InternalError("could not create version node")
-		}
-		defer file.Close()
-
-		fi, err := file.Stat()
-		if err != nil {
-			discardBlob()
-			sublog.Err(err).Str("version", versionsPath).Msg("could not stat version node")
-			return errtypes.InternalError("could not stat version node")
-		}
-		newMtime = fi.ModTime()
-
-		// copy blob metadata to version node
-		err = xattrs.CopyMetadataWithSourceLock(targetPath, versionsPath, func(attributeName string) bool {
-			return strings.HasPrefix(attributeName, xattrs.ChecksumPrefix) ||
-				attributeName == xattrs.BlobIDAttr ||
-				attributeName == xattrs.BlobsizeAttr
-		}, lock)
```

Comment:
we copy only the necessary blob metadata to the version. shares always stay in the ectual node. this got lost as well

Reply:
Ah. Thats new. I still use the old code https://github.com/cs3org/reva/blob/a4825d8b9ecc89a1635a9c536b7a578195cb3739/pkg/storage/utils/decomposedfs/upload/processing.go#L374-L386

So only Name, BlobID and BlobSize need to be copied?

Reply:
I'm creating only one lock now and pass it around. This will be much better performance wise